File: 4 - Athena's Enigma of the Hash Symbols/Solution.java
import java.util.*;

class Solution {
    public static String checkIfPairExists(String scroll){
        int hashCount = 0;
        Set<Integer> set = new HashSet<>();
        List<Integer> numbers = new ArrayList<>();
        for(int i = 0; i < scroll.length(); i++){
            char currentChar = scroll.charAt(i);
            if(currentChar == '#'){
                hashCount++;
            }
            else if(Character.isDigit(currentChar)){
                numbers.add(currentChar-'0');
            }

        }
        for(int i = 0; i < numbers.size(); i++){
            int key = hashCount - numbers.get(i);
            if(set.contains(key)){
                return "Yes";
            }
            else {
                set.add(numbers.get(i));
            }
        } 
        return "No";
}
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        while (N-- > 0) {
            String scroll = sc.next();
            System.out.println(checkIfPairExists(scroll));
        }
    }
}

File: 4 - Athena's Enigma of the Hash Symbols/solution.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

const char* checkIfPairExists(const char* scroll) {
    int hashCount = 0;
    int set[10] = {0}; // To store numbers (0-9)
    int numbers[100000];
    int numIndex = 0;

    for (int i = 0; scroll[i] != '\0'; i++) {
        if (scroll[i] == '#') {
            hashCount++;
        } else if (isdigit(scroll[i])) {
            numbers[numIndex++] = scroll[i] - '0';
        }
    }

    for (int i = 0; i < numIndex; i++) {
        int key = hashCount - numbers[i];
        if (key >= 0 && key < 10 && set[key]) {
            return "Yes";
        }
        set[numbers[i]] = 1;
    }

    return "No";
}

int main() {
    int N;
    scanf("%d", &N);
    while (N--) {
        char scroll[100001];
        scanf("%s", scroll);
        printf("%s\n", checkIfPairExists(scroll));
    }
    return 0;
}
